File: src/contest/w424/ZeroArrayTransformationII.java
/*
 * Copyright (c) 2024 THGEEK.COM. All rights reserved.
 */

package contest.w424;

/**
 * LeetCode 3356
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/11/17 10:53
 */
public class ZeroArrayTransformationII {
    public int minZeroArray(int[] nums, int[][] queries) {
        int sum = 0;
        for (int num : nums) {
            sum += num;
        }
        if (sum == 0) {
            return 0;
        }

        int k = 0;
        for (int[] query : queries) {
            int l = query[0], r = query[1];
            for (int i = l; i <= r; i++) {
                int decrement = Math.min(nums[i], query[2]);
                nums[i] -= decrement;
                sum -= decrement;
            }
            k++;
            if (sum == 0) {
                return k;
            }
        }
        return -1;
    }
}

File: src/contest/w424/ZeroArrayTransformationII.go
package w424

// LeetCode 3356
func minZeroArray(nums []int, queries [][]int) int {
	l := 0
	r := len(queries) + 1
	for l < r {
		mid := l + (r-l)/2
		if isZeroArray(nums, queries, mid) {
			r = mid
		} else {
			l = mid + 1
		}
	}

	if l <= len(queries) {
		return l
	} else {
		return -1
	}
}

func isZeroArray(nums []int, queries [][]int, k int) bool {
	line := make([]int, len(nums)+1)
	for i := 0; i < k; i++ {
		line[queries[i][0]] += queries[i][2]
		line[queries[i][1]+1] -= queries[i][2]
	}

	curr := 0
	for i := 0; i < len(nums); i++ {
		curr += line[i]
		if nums[i] > curr {
			return false
		}
	}
	return true
}
